pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/json/MutableJsonIndexImpl.java

```diff
@@ -106,8 +106,11 @@ private void addFlattenedRecords(List<Map<String, String>> records) {
         // Put both key and key-value into the posting list. Key is useful for checking if a key exists in the json.
         String key = entry.getKey();
         _postingListMap.computeIfAbsent(key, k -> new RoaringBitmap()).add(_nextFlattenedDocId);
-        String keyValue = key + JsonIndexCreator.KEY_VALUE_SEPARATOR + entry.getValue();
-        _postingListMap.computeIfAbsent(keyValue, k -> new RoaringBitmap()).add(_nextFlattenedDocId);
+        int length = _jsonIndexConfig.getMaxValueLength();
```

Comment:
It is already too late. It is very costly to flatten a huge record (can even run out of memory), so we should think of a limit that can be applied when flattening the JSON

Reply:
I see what you're saying. After discussing it with @chenboat and @itschrispeck, we'd like to truncate rather than omit long values from the index; I can move this logic into `flatten`. Additionally, we wanted to add a suffix to truncated values so that if/when we support group-by on JSON-indexed columns, the user can detect that a value was truncated.

I will update the PR description and the PR with the same. Let me know if you have major concerns with this design (and thanks for the review!).

Reply:
Truncating the value won't really work in the current JSON index because we can only do exact match.
Using the current json index storage format to support group-by is almost not possible. I discussed this with @yupeng9 and I believe the idea is to extract the column instead of using the json index to power group-by.

Reply:
+1
cc @chenboat

Reply:
@Jackie-Jiang Separating the truncation from group by issue - I think truncation will still be useful as a pre-filter or for approximation. In our case a superset would be preferable to missing docs

Reply:
@itschrispeck If we just want to keep the key so that value can still be retrieved by `json_extract_scalar`, we may remove the value (or put something like `__SKIPPED__`) and just keep the key. The only way to use the truncated value is to also query the truncated value, and I don't know if it worths the storage cost of storing it.

Reply:
@Jackie-Jiang This PR is orthogonal to the group by feature we brought earlier. Truncating the string length in json index itself is an improvement over the current Json index implementation.  String has maxLength limit in Pinot today but string in Json does not have that. This is the improvement this PR is about. A json string overall can be as long as 1M bytes in our system. We can not have string value in it to have such a length.

Reply:
@chenboat I totally understand the intention, and I agree we can make the longest allowed value configurable. I'm proposing saving even more space when the value is longer than the threshold by not storing the value at all because it is not queryable anyway.

Reply:
To confirm, we are proposing to truncate to a configurable max length of value like "very loooooooooooooooog value" in the json example below. It is not queryable because json index only supports exact match today? @Jackie-Jiang 
```
{
  ....
  "a": "very loooooooooooooooog value"
   ....
}
```

Reply:
Yes. In older to look up JSON index, you'll need to provide the exact value. JSON index also supports key only lookup. More details can be found [here](https://docs.pinot.apache.org/basics/indexing/json-index#how-to-use-the-json-index).
Based on this, I suggest replacing the very long value with a placeholder such as `"__SKIPPED__"` to save more space. This way we will get better compression as well because the key value pair might repeat thus can be compressed. The original value can still be retrieved by `JSON_EXTRACT_SCALAR` by looking up the key.

Reply:
* I moved the logic into `JsonUtils.flatten` and implemented the feature to replace long values with `"$SKIPPED$"`.
* I chose `"$SKIPPED"` to be similar to the constants in [`GenericRow`](https://github.com/apache/pinot/blob/master/pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java#L76), but I can change it to `__SKIPPED__` if that follows a convention in Pinot.
* I called the constant `SKIPPED_VALUE_REPLACEMENT` but another option might be `LONG_VALUE_REPLACEMENT`.

Let me know if you have any suggestions or concerns.